engine/execution/provider/engine_test.go

```diff
@@ -269,38 +313,43 @@ func TestProviderEngine_onChunkDataRequest(t *testing.T) {
 		blockID := unittest.IdentifierFixture()
 
 		execState.On("GetBlockIDByChunkID", chunkID).Return(blockID, nil)
-		ps.On("AtBlockID", blockID).Return(ss)
-
+		ps.On("AtBlockID", blockID).Return(ss).Once()
 		ss.On("Identity", originIdentity.NodeID).Return(originIdentity, nil).Once()
+
+		// channel tracking for the first chunk data pack request responded.
+		responded := make(chan struct{})
 		chunkConduit.On("Unicast", mock.Anything, originIdentity.NodeID).
 			Run(func(args mock.Arguments) {
 				res, ok := args[0].(*messages.ChunkDataResponse)
 				require.True(t, ok)
 
 				actualChunkID := res.ChunkDataPack.ChunkID
 				assert.Equal(t, chunkID, actualChunkID)
+				close(responded)
 			}).
 			Return(nil).Once()
 
-		execState.On("ChunkDataPackByChunkID", mock.Anything, chunkID).Return(chunkDataPack, nil).Twice()
+		execState.On("ChunkDataPackByChunkID", chunkID).Return(chunkDataPack, nil).Twice()
 
 		req := &messages.ChunkDataRequest{
 			ChunkID: chunkID,
 			Nonce:   rand.Uint64(),
 		}
 
+		cancelCtx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		ctx, _ := irrecoverable.WithSignaler(cancelCtx)
+		e.Start(ctx)
+		// submit using non-existing origin ID
 		unittest.RequireCloseBefore(t, e.Ready(), 100*time.Millisecond, "could not start engine")
+		require.NoError(t, e.Process(channels.RequestChunks, originIdentity.NodeID, req))
 
-		// an authorized request followed by an unauthorized one
-		e.onChunkDataRequest(context.Background(), originIdentity.NodeID, req)
+		// waits till first chunk data pack is replied and then makes the requester unauthorized.
+		unittest.RequireCloseBefore(t, responded, 1*time.Second, "could not get first chunk data pack responded")
```

Comment:
```suggestion
		// waits till request is ingested to a worker for processing
		assert.Eventually(t, func() bool { return chdpQueue.Size() == 0 }, time.Second, 10*time.Millisecond)
```
Suggesting to implement this synchronization in a way that will also prevent race conditions in the unhappy path cases. (concrete suggestion for previous comment)

* `chdpQueue` is the queue passed to the `provider.Engine` constructor

Reply:
Added the synchronization in the middle https://github.com/onflow/flow-go/pull/2951/commits/f32db517ed868b7636ee1aa577c2d2889d2e6bc9, removed this sync here: https://github.com/onflow/flow-go/pull/2951/commits/927d5399ca690d77387b1397899feeff402ce3b4